lib/FrontendTool/FrontendTool.cpp

```diff
@@ -1623,20 +1623,37 @@ int swift::performFrontend(ArrayRef<const char *> Args,
   // arguments are generated by the driver, not directly by a user. The driver
   // is responsible for emitting diagnostics for its own errors. See SR-2683
   // for details.
-  //
-  // FIXME: Do we need one SerializedConsumer per primary? Owned by the
-  // SourceFile?
-  std::unique_ptr<DiagnosticConsumer> SerializedConsumer;
+  std::unique_ptr<DiagnosticConsumer> SerializedConsumerMultiplexer;
   {
-    const std::string &SerializedDiagnosticsPath =
-        Invocation.getSerializedDiagnosticsPathForAtMostOnePrimary();
-    if (!SerializedDiagnosticsPath.empty()) {
-      SerializedConsumer.reset(
+    SmallVector<FileSpecificDiagnosticConsumer::ConsumerPair, 4>
```

Comment:
Why 4? There must be something about the implementation of `SmallVector`. If I had to guess, I would bet that the common cases are 1 and many, FWIW.

Reply:
After reading it, I see that the following code produces a vector of `ConsumerPairs`. Suggest adding either a comment to that effect or factoring into a subroutine. In any event, it would help to add a comment explaining why such a vector is needed.

Reply:
I think you're right about "1 and many", but there are lots of projects where "many" is still going to be <= 4. (On an eight-core machine, that supports up to 32 files in a clean build.) You're right that I just made the number up, though.

I don't understand what you mean about the following code. The vector is built up and then passed to the constructor that uses it. The code that does the building up says what it does. What would the subroutine be called?

(I could see factoring out the whole creation of the FileSpecificDiagnosticConsumer, including building this vector, but just the vector part doesn't seem to buy us anything.)

Reply:
Re: vector-building: Again, I'd be more inclined to pull out the whole multiplexer part (now "dispatcher" thanks to Pavel). I can do that. If I wanted to pin down why the vector computation isn't worth pulling out separately, I think it's because it's not reusable in any interesting way and it can only be described in terms of "this will then be used to create a FileSpecificDiagnosticConsumer".

Re: 4: I don't know. We have SmallVectors all over the place. Most of them have arbitrary initial sizes. 4 sounds good to me, but is it better than 8? I have no idea. I don't think it's worth my time to look into, but maybe it is.